devito/finite_differences/operations.py

```diff
@@ -0,0 +1,164 @@
+import sympy
+
+import devito
+from devito.symbolics.search import retrieve_functions
+from devito.symbolics.extended_sympy import FrozenExpr
+
+__all__ = ['Mul', 'Add', 'Pow']
+
+
+def fd_parameters(obj):
+    """
+    Process a set of Mul, Add, Pow expression and outputs
+    the space_order, time_order of the whole expression.
+    As the leading order is always the smallest order, if multiple
+    orders are found the lowest is chosen
+    """
+    func = list(retrieve_functions(obj))
+    time_dims = ()
+    sp_dims = ()
+    so = 100
+    to = 100
+    is_time = False
+    stagg = (0, 0, 0, 0)
+    # Filter expressions to get space and time Order
+    # The space order is the minimum space order
+    # The time order is the minimum time_order
+    for i in func:
+        if isinstance(i, devito.Function):
+            sp_dims += i.space_dimensions
+            so = min(so, i.space_order)
+            stagg = i.staggered
+        if isinstance(i, devito.TimeFunction):
+            sp_dims += i.space_dimensions
+            time_dims += (i.time_dim,)
+            so = min(so, i.space_order)
+            is_time = True
+            to = min(to, i.time_order)
+            stagg = i.staggered
+
+    obj.space_order = so
+    obj.time_order = to
+    obj.space_dimensions = tuple(set(sp_dims))
+    obj.staggered = stagg
+    if is_time:
+        obj.time_dim = tuple(set(time_dims))[0]
+    obj.is_TimeFunction = is_time
+
+
+class Pow(sympy.Mul, FrozenExpr):
+    """A customized version of :class:`sympy.Add` representing a sum of
+    symbolic object."""
+    def __new__(cls, *args, **kwargs):
+        return sympy.Pow.__new__(cls, *args)
+
+    def __init__(self, *args, **kwargs):
+        fd_parameters(self)
+        devito.finite_differences.finite_difference.initialize_derivatives(self)
+
+    def __add__(self, other):
+        return Add(self, other)
+
+    def __iadd__(self, other):
+        return Add(self, other)
+
+    def __radd__(self, other):
+        return Add(self, other)
+
+    def __mul__(self, other):
+        return Mul(self, other)
+
+    def __imul__(self, other):
+        return Mul(self, other)
+
+    def __rmul__(self, other):
+        return Mul(self, other)
+
+    @property
+    def _time_size(self):
+        func = list(retrieve_functions(self))
+        for i in func:
+            if isinstance(i, devito.TimeFunction):
+                return i._time_size
+
+        return None
+
+
+class Mul(sympy.Mul, FrozenExpr):
+    """A customized version of :class:`sympy.Add` representing a sum of
+    symbolic object."""
+    is_Mul = True
+
+    def __new__(cls, *args, **kwargs):
+        return sympy.Mul.__new__(cls, *args)
+
+    def __init__(self, *args, **kwargs):
+        fd_parameters(self)
+        devito.finite_differences.finite_difference.initialize_derivatives(self)
+
+    def __add__(self, other):
```

Comment:
are these __add__ etc really necessary? anyway, if so, you could put all these methods in the superclass I was mentioning earlier

Reply:
Yes they are for FD. Maily `u.dx` becomes an `Add` object and needs to know how to`_add__` itself to erform `u.dx.dy` for example. Same for Mul (FD coeff) and Pow (1/hx)

Reply:
no I meant the fact that they are identical across Pow/Add/Mul, so they should be moved to a common superclass once and for all

Reply:
Oh right, would be smarter yes